commit 562caf202d6c7293f641dc0d9012a119c01ea799 (refs/changes/96/1796/1)
Author: Marsell Kukuljevic <marsell@joyent.com>
Date:   2017-04-17T19:26:15+12:00 (2 years, 6 months ago)
    
    CNAPI-706: Add a command in bin/ which shows the reasoning behind dapi's most recent allocation

diff --git a/bin/alloc-reasons.sh b/bin/alloc-reasons.sh
new file mode 100755
index 0000000..4b861f4
--- /dev/null
+++ b/bin/alloc-reasons.sh
@@ -0,0 +1,26 @@
+#!/bin/bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Copyright (c) 2017, Joyent, Inc.
+#
+# This command shows the reasoning that dapi (embedded within cnapi)
+# used when deciding where the most recent allocation should go. When
+# dapi makes an allocation, it dumps a gzipped & base64-encoded JSON
+# blob into cnapi's logs which includes a great deal of information about
+# the allocation, including the 'steps' attribute; that attribute is what
+# we seek to extract here.
+
+set -o errexit
+
+# first dig up the most recent snapshot out of the logs
+grep -h snapshot $(svcs -L cnapi) $(ls /var/log/sdc/upload/cnapi_* 2> /dev/null | sort -r) \
+    | bunyan -c this.snapshot -o bunyan --strict \
+    | tail -1 \
+    | json -ga snapshot \
+| while read snap; do
+    # and now decode it so we can reach the 'steps' attribute
+    echo "$snap" | base64 -d | gunzip - | json steps
+done
